OpenUtau.Plugin.Builtin/AdvancedPhonemizer.cs

```diff
@@ -0,0 +1,520 @@
+        /// reads from dictionary if provided
+        /// </summary>
+        /// <param name="note"></param>
+        /// <returns></returns>
+        protected virtual string[] GetSymbols(Note note) {
+            string[] getSymbolsRaw(string lyrics) {
+                if (lyrics == null) {
+                    return new string[0];
+                } else return lyrics.Split(" ");
+            }
+
+            if (hasDictionary) {
+                if (!string.IsNullOrEmpty(note.phoneticHint)) {
+                    return getSymbolsRaw(note.phoneticHint);
+                }
+                var result = new List<string>();
+                foreach (var subword in note.lyric.Trim().ToLowerInvariant().Split(new string[] { " ", "_" }, StringSplitOptions.RemoveEmptyEntries)) {
+                    var subResult = dictionary.Query(subword);
+                    if (subResult == null) {
+                        subResult = HandleWordNotFound(subword);
+                        if (subword == null) {
+                            return null;
+                        }
+                    }
+                    result.AddRange(subResult);
+                }
+                return result.ToArray();
+            }
+            else {
+                return getSymbolsRaw(note.lyric);
+            }
+        }
+
+        /// <summary>
+        /// Checks if mapped and validated alias exists in oto
+        /// </summary>
+        /// <param name="alias"></param>
+        /// <param name="note"></param>
+        /// <returns></returns>
+        protected bool HasOto(string alias, int tone) {
+            return singer.TryGetMappedOto(ValidateAlias(alias), tone, out _);
+        }
+
+        /// <summary>
+        /// Instead of changing symbols in cmudict itself for each reclist, 
+        /// you may leave it be and provide symbol replacements with this method.
+        /// </summary>
+        /// <returns></returns>
+        protected virtual Dictionary<string, string> GetDictionaryPhonemesReplacement() { return null; }
+
+        /// <summary>
+        /// separates symbols to syllables and ending.
+        /// Not sure you would like to override it, but, anyway.
+        /// </summary>
+        /// <param name="notes"></param>
+        /// <param name="prevWord"></param>
+        /// <returns></returns>
+        protected virtual Word? MakeWord(Note[] notes, Word? prevWord = null) {
+            var mainNote = notes[0];
+            var symbols = GetSymbols(mainNote);
+            if (symbols == null) {
+                return null;
+            }
+            if (symbols.Length == 0) {
+                symbols = new string[] { "" };
+            }
+            symbols = ApplyExtensions(symbols, notes);
+            List<int> vowelIds = ExtractVowels(symbols);
+            if (vowelIds.Count == 0) {
+                // no syllables or all consonants, the last phoneme will be interpreted as vowel
+                vowelIds.Add(symbols.Length - 1);
```

Comment:
diphtongs contain main syllable-vowel and non-syllable vowel, like, "like" i is [aI] or [a y], depending of the reclist. When "like" is sang on a long note, only [a] part is stretched. In any manner it makes one syllable. If it's represented as [a y], [y] must be set as consonant. If it's [aI], then CV basically sounds like [l a], and vc sounds like [a y k].
"Vowel" in synth is just that sound that will be stretched. It may actually be a consonant, like "M" or "N".
i'll add a comment

Reply:
"like" `l ay k` in arpabet really is one vowel. "ours" `aw er s` has two vowels. But it does seem phoneme symbol set dependent. I'm fine with this for now.